### PR TITLE
Allow _Float16 as alias for float16_t in halide_type_of<>() (#7325)

### PR DESCRIPTION
**Pull request by steven-johnson**

### CODE REVIEW

File: test/correctness/float16_t.cpp [hunk 1]
```diff
@@ -290,43 +307,71 @@ int main(int argc, char **argv) {
...
         float16_t f(1.0f16);
         _Float16 f2 = (_Float16)f;
         if (f2 != 1.0f16) {
-            printf("Roundtrip of 16-bit float via _Float16 failed.\n");
+            fprintf(stderr, "Roundtrip of 16-bit float via _Float16 failed.\n");
             return -1;
         }
     }
+#else
+    printf("Only clang supports _Float16 literals, skipping roundtrip test\n");
```
Comment (terryheo) [inline test/correctness/float16_t.cpp:370]: I can't get this. In GCC, we can't use `_Float16` even if `HALIDE_CPP_COMPILER_HAS_FLOAT16` is set?

Comment (steven-johnson) [reply test/correctness/float16_t.cpp:370]: As far as I can tell, even the versions of gcc that support `_Float16` don't support 16-bit-floating-point-constants via the `f16` suffix. (Maybe I don't have it configured correctly, but I also can't find any gcc docs that suggest that it does.)

Comment (terryheo) [reply test/correctness/float16_t.cpp:370]: Got it. Thanks for the explanation.

Comment (terryheo) [inline test/correctness/float16_t.cpp:370]: It wasn't sufficient to me. Maybe
`Only clang supports _Float16 constant literal 'f16', , skipping roundtrip test` ?

File: src/Float16.h [hunk 2]
```diff
@@ -71,7 +58,9 @@ struct float16_t {
 #ifdef HALIDE_CPP_COMPILER_HAS_FLOAT16
     /** Cast to compiler's built-in _Float16 type. */
     explicit operator _Float16() const {
-        return *(const _Float16 *)&data;
+        _Float16 result;
```
Comment (terryheo) [inline src/Float16.h:61]: Just wonder why this change is needed. The old logic doesn't work?

**Review by terryheo — APPROVED** [review]
LGTM

File: src/Float16.h (hunk 2 above)
Comment (steven-johnson) [reply src/Float16.h:61]: The old logic was technically a violation of strict aliasing rules; in practice, ~all C++ compilers would work fine with it, but one of ours complained, and using memcpy() is definitely safe (and likely to be optimized into the same code).

File: test/correctness/float16_t.cpp (hunk 1 above)
Comment (steven-johnson) [reply test/correctness/float16_t.cpp]: Done

**Comment by steven-johnson** [comment]
Ping to @abadams and/or @zvookin

**Review by abadams — APPROVED** [review]

**Comment by steven-johnson** [comment]
I'm going to go ahead and land this to unblock some work; hopefully @zvookin can chime in when he has a chance, since he spent a lot of time on the original issue here.
